商品購入機能

miracletea opened the pull request:
# What
商品購入機能の作成

# Why
商品購入機能の実装のため

【Gyazo】
・必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
 https://gyazo.com/edd7dd8e55ad7a9aa37fd9cb28ba16f6

・入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/08c5e4e041e737eb7852690ef4896401

・ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/083740239f705e0bfcb71e8d9726cebc

・ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://gyazo.com/eda123e8c17e04aa729009c219c325ea

・ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/3ca11eabe7093718a0909b309dfcce9e

・売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
・ 売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
・ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
上記3項目共通
https://gyazo.com/58cfa5ab76a849781c2728f3782dc000


・ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/98763a64745ffb50a436d4a03903f380

・テスト結果の画像
https://gyazo.com/b86611e16e4b9bed1bdb11e9f5312dfb

miracletea committed:
oreder_formの作成、およびER図の変更

・order_formはDBにデータが保存されることを確認
・recordテーブルをorderテーブルに変更

miracletea committed:
order_formのエラー表示

miracletea committed:
order_formにテストコード作成

miracletea committed:
card機能(payjp)の作成

miracletea committed:
セミコロンの付け忘れ

miracletea committed:
画像上の「sold out」、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが非表示

miracletea committed:
javascriptからのデータ受け取り時にディレイを追加

miracletea committed:
商品が購入されてないときの編集画面遷移の制限

yuri240 commented on app/models/order_form.rb:
```diff
@@ -0,0 +1,30 @@
+  validates :ship_from_to_id, presence: true
+  validates :city_town_village, presence: true
+  validates :street_address, presence: true
+  validates :building_name, presence: false
```
```ruby
  validates :building_name, presence: false
```
建物名はバリデーションを記述する必要がないため、削除しておきましょう！

yuri240 commented on app/models/order_form.rb:
```diff
@@ -0,0 +1,30 @@
+  include ActiveModel::Model
+  attr_accessor :token, :user_id, :item_id, :postal_code, :ship_from_to_id, :city_town_village, :street_address, :building_name, :phone_number
+
+  validates :user_id, presence: true
```
`presence: true`は「with_options」を使用することで、
ひとまとまりにすることが可能となります。

つまり、
1回の記述ですべてのカラムに`presence: true`が適応されるよう修正しましょう！

▼以下に参考カリキュラムをご案内させていただきます。
https://master.tech-camp.in/v2/curriculums/6771
こちらの「`with_options（ウィズ オプションズ）`」をご確認ください。

yuri240 commented on spec/factories/order_forms.rb:
```diff
@@ -0,0 +1,14 @@
+  factory :order_form do
+    Faker::Config.locale = :ja
+    user_id { Faker::Number.non_zero_digit }
+    item_id { Faker::Number.non_zero_digit}
```
```ruby
    user_id { Faker::Number.non_zero_digit }
    item_id { Faker::Number.non_zero_digit}
```
外部キーに直接数字を入れるのは正しい情報が入っているかが分からないため、好ましい方法ではありません。

specファイルにて、値を生成する形にしましょう。

該当ファイル(`spec/models/order_form_spec.rb`)にて、記述例を上げさせていただきます。

yuri240 commented on spec/models/order_form_spec.rb:
```diff
@@ -0,0 +1,85 @@
+RSpec.describe OrderForm, type: :model do
+before do
+  @order_form = FactoryBot.build(:order_form)
+end
```
user_id、item_idはbeforeブロック内で生成しましょう。

以下は記述例となります。
```ruby
before do
 ユーザーid = FactoryBot.create(:user)
 商品id = FactoryBot.create(:item)
 @order_form = FactoryBot.build(:order_form, user_id: ユーザーid, item_id: 商品id)
end
```

▼以下に参考カリキュラムをご案内させていただきます。
https://master.tech-camp.in/v2/curriculums/6771
こちらの「`Formオブジェクトパターンで実装した場合のテストコードの書き方`」をご確認ください。

yuri240 commented on spec/models/order_form_spec.rb:
```diff
@@ -0,0 +1,85 @@
+        @order_form.valid?
+        expect(@order_form.errors.full_messages).to include("Postal code is invalid")
+      end
+      it '電話番号(phone_number)は、10桁以上11桁以内の半角数値、以外は登録できない' do
```
`電話番号(phone_number)は、10桁以上11桁以内の半角数値、以外は登録できない`

 上記のテストは見本のフリマアプリに合わせて、以下テスト項目に修正をお願いいたします。
```
・9桁以下では登録できないこと
・12桁以上では登録できないこと
```

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
コメントさせていただきましたので、ご確認お願いいたします。

また、お手数ではありますがテストコードを追加した後は、
改めてテストコード実行後のキャプチャ画像を撮影いただき、
こちらのプルリクエストへ添付していただくようお願いいたします。

miracletea committed:
建物名のバリデーション削除

app/models/order_form.rb
  validates :ship_from_to_id, presence: true
  validates :city_town_village, presence: true
  validates :street_address, presence: true
  validates :building_name, presence: false
@yuri240 yuri240 19 minutes ago
  validates :building_name, presence: false
建物名はバリデーションを記述する必要がないため、削除しておきましょう！

miracletea committed:
presence:trueをwith_optionに変更

miracletea committed:
user_id、item_idはブロック内で生成、およびjavascriptのため非同期に設定

spec/models/order_form_spec.rb
Comment on lines +4 to +6
before do
  @order_form = FactoryBot.build(:order_form)
end
@yuri240 yuri240 3 hours ago
user_id、item_idはbeforeブロック内で生成しましょう。

以下は記述例となります。

before do
 ユーザーid = FactoryBot.create(:user)
 商品id = FactoryBot.create(:item)
 @order_form = FactoryBot.build(:order_form, user_id: ユーザーid, item_id: 商品id)
end
▼以下に参考カリキュラムをご案内させていただきます。
https://master.tech-camp.in/v2/curriculums/6771
こちらの「Formオブジェクトパターンで実装した場合のテストコードの書き方」をご確認ください。

miracletea committed:
電話番号のテストを2つに分けて記述

spec/models/order_form_spec.rb
        @order_form.valid?
        expect(@order_form.errors.full_messages).to include("Postal code is invalid")
      end
      it '電話番号(phone_number)は、10桁以上11桁以内の半角数値、以外は登録できない' do
@yuri240 yuri240 3 hours ago
電話番号(phone_number)は、10桁以上11桁以内の半角数値、以外は登録できない

 上記のテストは見本のフリマアプリに合わせて、以下テスト項目に修正をお願いいたします。

・9桁以下では登録できないこと
・12桁以上では登録できないこと

miracletea commented:
ご確認ありがとうございます。
コメント内容を反映しました、宜しくお願い致します。
また、ご教示頂いたカリキュラムはエラーになり観覧できなかったため、ご容赦ください。
リンクを開くと以下が表示せれる。
「閲覧できないカリキュラム、または受講期間が終了しているためカリキュラムを表示できません」

【Gyazo】
テスト結果
https://gyazo.com/8ab6989299bf322c284db5c0dab32d17

TakashiKozuma commented:
レビュー依頼ありがとうございます。
LGTMです！実装お疲れさまでした。

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に次の実装のためのブランチを切ってから次にお進み下さい。
